src/state.rs

```diff
@@ -103,13 +110,45 @@ impl State {
         } else {
             self.board.move_piece(from, to)
         };
+        let next_castling = self.castling.after_move(self.player, from);
         let next_state = State {
             board: next_board,
             player: self.player.other(),
             en_passant: self.en_passant_pos(from, to),
+            castling: next_castling,
         };
         Move {
-            index: (from, to),
+            action: Action::Simple { from, to },
+            next: next_state,
+        }
+    }
+
+    fn can_castle_without_check(&self, kingside: bool) -> bool {
```

Comment:
It would be nice to indicate to readers that this logic is only ever used from `can_castle`. Possible ways that this could be done:

- Inline the fn
- Make the fn a closure inside of `can_castle`
- Put them both in a mod and make only `can_castle` pub

Reply:
yeah, this should probably all be one function. the reason i split it into two was that i wanted a short-circuit: if a player had already castled, i don't want to build up potential king positions to check if they can castle, etc. but maybe it's better with just an early return clause: 9b7ba60d11c1edef880f6461da3a802b62a12038